src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Abstractions/Contracts/IThread.cs

```diff
@@ -0,0 +1,56 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.Diagnostics.DataContractReader.Contracts;
+
+internal record struct ThreadStoreData(
```

Comment:
Perhaps a plain ol' struct would be better? I'm not sure how much it has improved since https://github.com/dotnet/linker/issues/1924, but for the record, the track record of C# records isn't great when it comes to codegen 😁

Reply:
For public APIs, you're correct. For internal implementation details and this use case in particular, I think it is appropriate.

Reply:
(note: I didn't modify this bit - just moved it to a different file)

This particular `ThreadStoreData` _is_ part of a "public" API - its the return value of `IThread.GetThreadStoreData` . Currently internal because we're not committing to a public API for the cdac contracts yet, but if we did, this would certainly be part of it.

If we would rather make this a non-`record` struct, I'd like to make that change separately from this PR.  We can decide now and I'll do it in a follow-up.

Reply:
Converting this to a `record` isn't necessarily appropriate at this point. Something to consider going forward, but we need a concrete example for why it is a concern with the cDAC.

Reply:
> Converting this to a `record` isn't necessarily appropriate at this point.

Proposal was to convert it to `struct`, whose [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8BLAOwxilOwBsACXDKAVzAzoBUALWbAEwDKGaDADCEZuVwAKALAAoOkrpl2AVVKNsUCn268+4yRjQLlK8nQBC2MAGsA5lAmk9PGPyPlTi5aroACjCuZA76HoYuJmZ+lgAiEeGeUQCUANxAA) is leaner than [that of `record`'s](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8BLAOwxilOwBsACWSKAEztwygFcwM6AVABaxszAMoZoMAMIRO5XAAoAsACg66umV4BVUu2xQKzQcOYy5GNKo2bydAELYwAawDmUWaWNCYI8+Ss1DS06AAUYLzJXE18zT0trYLsAEViYv3iASgBuIA==). Unless we are using all features of record (overridden operators, ToString, Equals, Construct/Deconstruct etc.), it is not recommended to opt into it to save like two lines of C#. Today, linker can't shave all that off either.

Reply:
Is the cDAC a trimmer concern at present? I understand the implications here and it doesn't seem to rise to a concern here. I agree the use of `record` in a public API is not recommended and something to avoid. I also understand that extensive use of them across a component that is expected to be trimmed can be problematic. I do not think it is appropriate to apply this logic in a categorical fashion though. Several compoents use `record` types without any issue as internal implmentation details and that has been the case for some time - see the `LibraryImport` and COM interop source generators.

Reply:
LibraryImport's usage of record is the only reason why sr/libraries/{non-test} has some usages of `record` showing up lately. It is arguably "unnecessary". Very rarely libraries team does not opt into new roslyn features, and `record` data structure was one of those rare features which was not bought, and for a good reason "why would I do that? to save a line of code or two? no thanks!". :)

Also, since the primary constructor feature, if anything `record` is +1 keyword which injects extra helpers that we rarely need (if at all!):

```diff
- internal record struct ThreadStoreCounts(
+ internal struct ThreadStoreCounts(
    int UnstartedThreadCount,
    int BackgroundThreadCount,
    int PendingThreadCount,
    int DeadThreadCount);
```

If there is no real benefit of preferring `record` over `class`/`struct` in LibraryImport or cDAC, it's best to avoid it IMHO.